src/crystal/system/unix/getrandom.cr

```diff
@@ -1,116 +1,48 @@
-{% skip_file unless flag?(:linux) %}
-
-require "c/unistd"
-require "./syscall"
-
-{% if flag?(:interpreted) %}
-  lib LibC
-    fun getrandom(buf : Void*, buflen : SizeT, flags : UInt32) : LibC::SSizeT
-  end
-
-  module Crystal::System::Syscall
-    GRND_NONBLOCK = 1u32
-
-    # TODO: Implement syscall for interpreter
-    def self.getrandom(buf : UInt8*, buflen : LibC::SizeT, flags : UInt32) : LibC::SSizeT
-      LibC.getrandom(buf, buflen, flags)
+require "c/sys/random"
+
+{% if !LibC.has_method?(:getrandom) %}
+  # getrandom isn't available in android before API LEVEL 28
+  require "./urandom"
+{% else %}
+  module Crystal::System::Random
```

Comment:
**polish:** Putting almost the entire file content in a macro if isn't very neat. We can use `skip_file` instead.
```suggestion
  {% skip_file %}
{% end %}
module Crystal::System::Random
```

Reply:
That doesn't work, crystal complains about missing methods.

Reply:
```
$ ANDROID_NATIVE_API_LEVEL=android-24 bin/crystal build hello.cr --target aarch64-linux-android --cross-compile --no-codegen --progress --error-trace
In src/crystal/hasher.cr:83:27

 83 | Crystal::System::Random.random_bytes(@@seed.to_slice.to_unsafe_bytes)
                              ^-----------
Error: instantiating 'Crystal::System::Random.random_bytes(Slice(UInt8))'


In src/crystal/system/unix/urandom.cr:17:17

 17 | init unless @@initialized
                  ^
Error: can't infer the type of class variable '@@initialized' of Crystal::System::Random

Could you add a type annotation like this

    module Crystal::System::Random
      @@initialized : Type
    end

replacing `Type` with the expected type of `@@initialized`?

```

Reply:
This seems like very weird compiler behaviour. We should look into that.
But I found a working solution with moving the macro logic into `crystal/system/random.cr` where it actually fits much better.

```patch
diff --git c/src/crystal/system/random.cr i/src/crystal/system/random.cr
index 1a5b3c8f4..9815a04d1 100644
--- c/src/crystal/system/random.cr
+++ i/src/crystal/system/random.cr
@@ -13,7 +13,12 @@ end
 {% if flag?(:wasi) %}
   require "./wasi/random"
 {% elsif flag?(:linux) %}
-  require "./unix/getrandom"
+  {% if LibC.has_method?(:getrandom) %}
+    require "./unix/getrandom"
+  {% else %}
+    # getrandom isn't available in android before API LEVEL 28
+    require "./unix/urandom"
+  {% end %}
 {% elsif flag?(:bsd) || flag?(:darwin) %}
   require "./unix/arc4random"
 {% elsif flag?(:unix) %}
diff --git c/src/crystal/system/unix/getrandom.cr i/src/crystal/system/unix/getrandom.cr
index 0109fc773..6ad217c7c 100644
--- c/src/crystal/system/unix/getrandom.cr
+++ i/src/crystal/system/unix/getrandom.cr
@@ -1,44 +1,39 @@
 require "c/sys/random"
 
-{% if !LibC.has_method?(:getrandom) %}
-  # getrandom isn't available in android before API LEVEL 28
-  require "./urandom"
-{% else %}
-  module Crystal::System::Random
-    # Reads n random bytes using the Linux `getrandom(2)` syscall.
-    def self.random_bytes(buffer : Bytes) : Nil
-      getrandom(buffer)
-    end
+module Crystal::System::Random
+  # Reads n random bytes using the Linux `getrandom(2)` syscall.
+  def self.random_bytes(buffer : Bytes) : Nil
+    getrandom(buffer)
+  end
 
-    def self.next_u : UInt8
-      buffer = uninitialized UInt8
-      getrandom(pointerof(buffer).to_slice(1))
-      buffer
-    end
+  def self.next_u : UInt8
+    buffer = uninitialized UInt8
+    getrandom(pointerof(buffer).to_slice(1))
+    buffer
+  end
 
-    # Reads n random bytes using the Linux `getrandom(2)` syscall.
-    private def self.getrandom(buffer)
-      # getrandom(2) may only read up to 256 bytes at once without being
-      # interrupted or returning early
-      chunk_size = 256
+  # Reads n random bytes using the Linux `getrandom(2)` syscall.
+  private def self.getrandom(buffer)
+    # getrandom(2) may only read up to 256 bytes at once without being
+    # interrupted or returning early
+    chunk_size = 256
 
-      while buffer.size > 0
-        read_bytes = 0
+    while buffer.size > 0
+      read_bytes = 0
 
-        loop do
-          # pass GRND_NONBLOCK flag so that it fails with EAGAIN if the requested
-          # entropy was not available
-          read_bytes = LibC.getrandom(buffer, buffer.size.clamp(..chunk_size), LibC::GRND_NONBLOCK)
-          break unless read_bytes == -1
+      loop do
+        # pass GRND_NONBLOCK flag so that it fails with EAGAIN if the requested
+        # entropy was not available
+        read_bytes = LibC.getrandom(buffer, buffer.size.clamp(..chunk_size), LibC::GRND_NONBLOCK)
+        break unless read_bytes == -1
 
-          err = Errno.value
-          raise RuntimeError.from_os_error("getrandom", err) unless err.in?(Errno::EINTR, Errno::EAGAIN)
+        err = Errno.value
+        raise RuntimeError.from_os_error("getrandom", err) unless err.in?(Errno::EINTR, Errno::EAGAIN)
 
-          ::Fiber.yield
-        end
-
-        buffer += read_bytes
+        ::Fiber.yield
       end
+
+      buffer += read_bytes
     end
   end
-{% end %}
+end
diff --git c/src/crystal/system/unix/urandom.cr i/src/crystal/system/unix/urandom.cr
index fe81129a8..e05a85695 100644
--- c/src/crystal/system/unix/urandom.cr
+++ i/src/crystal/system/unix/urandom.cr
@@ -1,3 +1,5 @@
+require "c/sys/random"
+
 module Crystal::System::Random
   @@initialized = false
   @@urandom : ::File?
```

Reply:
@straight-shoota it compiles... but it always selects `urandom` instead of `getrandom` :sob:

Reply:
What... how? 😲

Reply:
I found a solution in the last commit. Require c/sys/random and delay processing the macros so we can load the correct file based on the presence of `LibC.getrandom`.

Reply:
@straight-shoota because it loaded c/sys/random _after_ processing the macros, so the libc method isn't defined.

Reply:
Ah yeah, makes sense. 👍 
I'm wondering if it makes sense to move this macro to a separate file then (`system/unix/random`) to make it a bit cleaner? Probably doesn't make much difference either way though.